content/blog/2025-12-18-extending-sql.md

```diff
@@ -0,0 +1,379 @@
+
+- Non-standard operators (JSON / geometry / regex operators)
+- Custom function syntaxes
+- Special identifier behavior
+
+### Example: Postgres JSON operators (`->`, `->>`)
+
+The Postgres `->` operator is a good illustration because it is widely used and parses only under the PostgreSQL dialect.
+
+Configure the dialect:
+
+```rust
+let config = SessionConfig::new()
+    .set_str("datafusion.sql_parser.dialect", "postgres");
+let ctx = SessionContext::new_with_config(config);
+```
+
+Then implement `ExprPlanner` to map the parsed operator (`BinaryOperator::Arrow`) to DataFusion semantics:
+
+```rust
+fn plan_binary_op(&self, expr: RawBinaryExpr, _schema: &DFSchema)
+  -> Result<PlannerResult<RawBinaryExpr>> {
+  match expr.op {
+    BinaryOperator::Arrow => Ok(Planned(/* your Expr */)),
+    _ => Ok(Original(expr)),
+  }
+}
+```
+
+Return `Planned(...)` when you handled the expression; return `Original(...)` to pass it to the next planner.
+
+For a complete JSON implementation, see [datafusion-functions-json]. For a minimal end-to-end example in the DataFusion repo, see [`expr_planner_tests`][expr_planner_tests].
+
+---
+
+## 3) Extending type support: `TypePlanner`
+
+After expressions, types are often the next thing to break. Schemas and DDL may reference types that DataFusion does not support out of the box, like MySQL's `DATETIME`.
+
+Type planning tends to come up when interoperating with other systems. You want to accept DDL or infer schemas from external catalogs without forcing users to rewrite types.
+
+`TypePlanner` maps SQL types to Arrow/DataFusion types:
+
+```rust
+impl TypePlanner for MyTypePlanner {
+  fn plan_type(&self, sql_type: &ast::DataType) -> Result<Option<DataType>> {
+    match sql_type {
+      ast::DataType::Datetime(Some(3)) => Ok(Some(DataType::Timestamp(TimeUnit::Millisecond, None))),
+      _ => Ok(None), // let the default planner handle it
+    }
+  }
+}
+```
+
+It is installed when building session state:
+
+```rust
+let state = SessionStateBuilder::new()
+  .with_default_features()
+  .with_type_planner(Arc::new(MyTypePlanner))
+  .build();
+```
+
+Once installed, if your `CREATE EXTERNAL CATALOG` statement exposes tables with MySQL types, DataFusion can interpret them correctly.
+
+---
+
+## 4) Extending the FROM clause: `RelationPlanner`
+
+Some extensions change what a _relation_ means, not just expressions or types. `RelationPlanner` intercepts FROM-clause constructs while SQL is being converted into a `LogicalPlan`.
+
+`RelationPlanner` originally came out of trying to build `MATCH_RECOGNIZE` support in DataFusion as a Datadog hackathon project. `MATCH_RECOGNIZE` is a complex SQL feature for detecting patterns in sequences of rows, and it made sense to prototype as an extension first. At the time, DataFusion had no extension point at the right stage of SQL-to-rel planning to intercept and reinterpret relations.
```

Comment:
I suggest considering moving these paragraphs  about the design history to after showing how it works (strategy A and strategy B sections) and putting it in its own sub section named something like like "Background" or "Origin of the API"

I think that would 
1. Make this section more consistent with the rest of the sections
2. Make it easier to quickly find the (great) examples here for people who are rushing